lib/spack/spack/environment.py

```diff
@@ -976,6 +989,67 @@ def remove(self, query_spec, list_name=user_speclist_name, force=False):
                 del self.concretized_order[i]
                 del self.specs_by_hash[dag_hash]
 
+    def develop(self, spec, path, clone=False):
+        """Add dev-build info for spec, set to version and path.
+
+        Returns (bool): True iff the environment was changed.
+        """
+        spec = spec.copy()  # defensive copy since we access cached attributes
+
+        if not spec.versions.concrete:
+            raise SpackEnvironmentError(
+                'Cannot develop spec %s without a concrete version' % spec)
+
+        for name, entry in self.dev_specs.items():
+            if name == spec.name:
+                e_spec = Spec(entry['spec'])
+                e_path = entry['path']
+
+                if e_spec == spec:
+                    if path == e_path:
+                        tty.msg("Spec %s already configured for development" %
+                                spec)
+                        return False
+                    else:
+                        tty.msg("Updating development path for spec %s" % spec)
+                        break
+                else:
+                    msg = "Updating development spec for package "
+                    msg += "%s with path %s" % (spec.name, path)
+                    tty.msg(msg)
+                    break
+        else:
+            tty.msg("Configuring spec %s for development at path %s" %
+                    (spec, path))
+
+        if clone:
+            # Not really cloning if it's a url-fetched version
+            abspath = path if os.path.isabs(path) else os.path.join(
+                self.path, path)
+
+            # Iterate over composite strategy to update path
+            # This will include resources in what we clone
+            for stage_obj in spec.package.stage:
+                stage_obj.path = abspath
+                stage_obj.source_path = abspath
```

Comment:
That's not all that it requires. The `Stage` puts everything in `spack-src`, but we don't want that for developer builds. So the stage has to know (every component of the composite stage has to know) to set `self.source_path` = `self.path`.

Reply:
That would cause you to lose the resources and hidden files, which are put in `stage.path` not `stage.source_path`. Not everything in the stage is in the source path.

Reply:
`ResourceStage.expand_archive` places its files into the root stage, so while it's true that you lose the resource base directories, I would assume they can be discarded. As I understand it the idea is to place everything needed for the installation into `Stage.source_path`. I'm not sure where hidden files would come from - do you have an example?

Reply:
I don't have an example, the idea only came to me because I was in the source code figuring out how to support this "flattened" use case